File: backend/app/schemas/analysis.py
"""
Pydantic schemas for analysis API
"""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# Base schemas for request/response
class BoundingBoxSchema(BaseModel):
    """Bounding box coordinates"""

    x: int
    y: int
    width: int
    height: int


class DetectedElementSchema(BaseModel):
    """Schema for a detected element"""

    bounding_box: BoundingBoxSchema
    confidence: float = Field(ge=0.0, le=1.0)
    label: str | None = None
    element_type: str | None = None
    screenshot_index: int = 0
    metadata: dict[str, Any] = Field(default_factory=dict)


class AnalyzerResultSchema(BaseModel):
    """Schema for results from a single analyzer"""

    analyzer_type: str
    analyzer_name: str
    elements: list[DetectedElementSchema]
    confidence: float
    metadata: dict[str, Any] = Field(default_factory=dict)


class FusedElementSchema(BaseModel):
    """Schema for a fused element from multiple analyzers"""

    bounding_box: BoundingBoxSchema
    confidence: float
    sources: list[str]
    source_confidences: dict[str, float]
    votes: int
    label: str | None = None
    element_type: str | None = None
    screenshot_index: int = 0
    metadata: dict[str, Any] = Field(default_factory=dict)


class AnalyzerInfoSchema(BaseModel):
    """Information about an available analyzer"""

    name: str
    type: str
    version: str
    supports_multi_screenshot: bool
    required_screenshots: int
    default_parameters: dict[str, Any]


# Request schemas
class AnalysisRequest(BaseModel):
    """Request to run analysis on an annotation set"""

    annotation_set_id: UUID
    analyzer_names: list[str] | None = None  # None = all analyzers
    analyzer_configs: dict[str, dict[str, Any]] | None = None
    parallel: bool = True
    fuse_results: bool = True
    overlap_threshold: float = Field(default=0.5, ge=0.0, le=1.0)
    save_to_database: bool = True  # Whether to save results to DB


class QuickAnalysisRequest(BaseModel):
    """Quick analysis request (no DB storage)"""

    annotation_set_id: UUID
    analyzers: list[str] | None = None
    fuse_results: bool = True


# Response schemas
class AnalysisResponse(BaseModel):
    """Response from running analysis"""

    analysis_job_id: UUID | None = None  # None if not saved to DB
    progress_job_id: UUID | None = None  # For polling progress (only during analysis)
    annotation_set_id: UUID
    analyzer_results: list[AnalyzerResultSchema]
    fused_elements: list[FusedElementSchema] | None = None
    fusion_stats: dict[str, Any] | None = None
    analyzer_statistics: dict[str, Any]
    status: str = "completed"
    error_message: str | None = None


class AnalysisJobSchema(BaseModel):
    """Schema for a stored analysis job"""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    annotation_set_id: UUID
    analyzers_used: list[str]
    parameters: dict[str, Any] | None = None
    fusion_enabled: bool
    fusion_config: dict[str, Any] | None = None
    status: str
    started_at: datetime | None = None
    completed_at: datetime | None = None
    error_message: str | None = None
    total_elements_found: int
    total_fused_elements: int
    analyzer_statistics: dict[str, Any] | None = None
    created_at: datetime
    created_by_id: UUID


class AnalysisJobDetailSchema(AnalysisJobSchema):
    """Detailed analysis job with elements"""

    fused_elements: list[FusedElementSchema]


class AnalysisJobListResponse(BaseModel):
    """Response for listing analysis jobs"""

    jobs: list[AnalysisJobSchema]
    total: int
    page: int
    page_size: int


# Utility schemas
class AnalyzerListResponse(BaseModel):
    """Response listing available analyzers"""

    analyzers: list[AnalyzerInfoSchema]
    total: int


class AnalysisStatistics(BaseModel):
    """Statistics about analysis results"""

    total_jobs: int
    total_elements_detected: int
    avg_confidence: float
    analyzers_used: dict[str, int]  # {analyzer_name: count}

File: frontend/src/services/analysis.ts
/**
 * Analysis API service
 */

import axios from "axios";

const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export interface BoundingBox {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface DetectedElement {
  bounding_box: BoundingBox;
  confidence: number;
  label?: string;
  element_type?: string;
  screenshot_index: number;
  metadata: Record<string, unknown>;
}

export interface AnalyzerResult {
  analyzer_type: string;
  analyzer_name: string;
  elements: DetectedElement[];
  confidence: number;
  metadata: Record<string, unknown>;
}

export interface FusedElement {
  bounding_box: BoundingBox;
  confidence: number;
  sources: string[];
  source_confidences: Record<string, number>;
  votes: number;
  label?: string;
  element_type?: string;
  screenshot_index: number;
  metadata: Record<string, unknown>;
}

export interface AnalyzerInfo {
  name: string;
  type: string;
  version: string;
  supports_multi_screenshot: boolean;
  required_screenshots: number;
  default_parameters: Record<string, unknown>;
}

export interface AnalysisResponse {
  analysis_job_id?: string;
  annotation_set_id: string;
  analyzer_results: AnalyzerResult[];
  fused_elements?: FusedElement[];
  fusion_stats?: {
    total_elements: number;
    avg_confidence: number;
    multi_vote_elements: number;
  };
  analyzer_statistics: Record<string, unknown>;
  status: string;
  error_message?: string;
}

export interface AnalysisJob {
  id: string;
  annotation_set_id: string;
  analyzers_used: string[];
  parameters?: Record<string, unknown>;
  fusion_enabled: boolean;
  fusion_config?: Record<string, unknown>;
  status: string;
  started_at?: string;
  completed_at?: string;
  error_message?: string;
  total_elements_found: number;
  total_fused_elements: number;
  analyzer_statistics?: Record<string, unknown>;
  created_at: string;
  created_by_id: string;
}

export interface AnalysisJobDetail extends AnalysisJob {
  fused_elements: FusedElement[];
}

export interface AnalysisRequest {
  annotation_set_id: string;
  analyzer_names?: string[];
  analyzer_configs?: Record<string, Record<string, unknown>>;
  parallel?: boolean;
  fuse_results?: boolean;
  overlap_threshold?: number;
  save_to_database?: boolean;
}

export interface QuickAnalysisRequest {
  annotation_set_id: string;
  analyzers?: string[];
  fuse_results?: boolean;
}

/**
 * Get list of available analyzers
 */
export async function listAnalyzers(token: string): Promise<AnalyzerInfo[]> {
  const response = await axios.get(`${API_URL}/api/v1/analysis/analyzers`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return response.data.analyzers;
}

/**
 * Run full analysis with database storage
 */
export async function runAnalysis(
  request: AnalysisRequest,
  token: string
): Promise<AnalysisResponse> {
  const response = await axios.post(
    `${API_URL}/api/v1/analysis/analyze`,
    request,
    {
      headers: { Authorization: `Bearer ${token}` },
    }
  );
  return response.data;
}

/**
 * Run quick analysis without database storage
 */
export async function runQuickAnalysis(
  request: QuickAnalysisRequest,
  token: string
): Promise<AnalysisResponse> {
  const response = await axios.post(
    `${API_URL}/api/v1/analysis/analyze/quick`,
    request,
    {
      headers: { Authorization: `Bearer ${token}` },
    }
  );
  return response.data;
}

/**
 * List analysis jobs
 */
export async function listAnalysisJobs(
  token: string,
  params?: {
    annotation_set_id?: string;
    status?: string;
    page?: number;
    page_size?: number;
  }
): Promise<{
  jobs: AnalysisJob[];
  total: number;
  page: number;
  page_size: number;
}> {
  const response = await axios.get(`${API_URL}/api/v1/analysis/jobs`, {
    params,
    headers: { Authorization: `Bearer ${token}` },
  });
  return response.data;
}

/**
 * Get detailed analysis job results
 */
export async function getAnalysisJob(
  jobId: string,
  token: string
): Promise<AnalysisJobDetail> {
  const response = await axios.get(`${API_URL}/api/v1/analysis/jobs/${jobId}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return response.data;
}

/**
 * Delete analysis job
 */
export async function deleteAnalysisJob(
  jobId: string,
  token: string
): Promise<void> {
  await axios.delete(`${API_URL}/api/v1/analysis/jobs/${jobId}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
}
